Added crypto_pwhash_scryptsalsa208sha256 support for key derivation.

terrorbyte opened the pull request:
Since this seems to be the unofficial python3 branch of pysodium I figured I'd push the scrypt key derivation to here. You will notice that the tests are very slow and CPU bound now, I can change the default scrypt parameters to be INTERACTIVE instead of SENSITIVE to speed it up if we need.

terrorbyte commented:
If we merge feel free to push a new pull request to the main fork.

bgaifullin commented on pysodium/__init__.py:
```diff
@@ -51,7 +51,13 @@
+crypto_pwhash_scryptsalsa208sha256_MEMLIMIT_INTERACTIVE = sodium.crypto_pwhash_scryptsalsa208sha256_memlimit_interactive()
+crypto_pwhash_scryptsalsa208sha256_OPSLIMIT_SENSITIVE = sodium.crypto_pwhash_scryptsalsa208sha256_opslimit_sensitive()
+crypto_pwhash_scryptsalsa208sha256_MEMLIMIT_SENSITIVE = sodium.crypto_pwhash_scryptsalsa208sha256_memlimit_sensitive()
+crypto_pwhash_scryptsalsa208sha256_SALTBYTES = sodium.crypto_pwhash_scryptsalsa208sha256_saltbytes()
```
crypto_pwhash_scryptsalsa208sha256_SALTBYTES - declared above.

bgaifullin commented on test/test_pysodium.py:
```diff
@@ -104,6 +104,15 @@ def test_crypto_stream_chacha20_xor(self):
+    def test_crypto_pwhash_scryptsalsa208sha256(self):
+        passwd = "howdy"
+        outlen = 128
+        salt = pysodium.randombytes(pysodium.crypto_pwhash_scryptsalsa208sha256_SALTBYTES);
```
extra symbol ';' at end of line

bgaifullin commented on pysodium/__init__.py:
```diff
@@ -150,6 +156,15 @@ def crypto_generichash_final(state, outlen=crypto_generichash_BYTES):
     return buf.raw
 
+# crypto_pwhash_scryptsalsa208sha256(unsigned char * const out, unsigned long long outlen, const char * const passwd, unsigned long long passwdlen, const unsigned char * const salt, unsigned long long opslimit,size_t memlimit);
+def crypto_pwhash_scryptsalsa208sha256(size, passwd, salt, opslimit = crypto_pwhash_scryptsalsa208sha256_OPSLIMIT_SENSITIVE, memlimit = crypto_pwhash_scryptsalsa208sha256_MEMLIMIT_SENSITIVE):
```
pep8: extra space between argname and default value

bgaifullin commented on pysodium/__init__.py:
```diff
@@ -150,6 +156,15 @@ def crypto_generichash_final(state, outlen=crypto_generichash_BYTES):
+    passwdlen = ctypes.c_ulonglong(len(passwd))
+    opslimit = ctypes.c_ulonglong(opslimit)
+    memlimit = ctypes.c_size_t(memlimit)
+    __check(sodium.crypto_pwhash_scryptsalsa208sha256(out, outlen, passwd, passwdlen, str(salt), opslimit, memlimit))
```
I think it is extra conversation to str for salt.

bgaifullin commented:
Thank you for you request. There is some comments from my-side.

terrorbyte commented:
Good catches. Bad habits from C die hard.

terrorbyte commented:
I removed the string conversion (automatic encoding of bytes is a bad idea) and fixed the test to have the correct output. Thanks for being so diligent checking that code.

bgaifullin commented:
Thank you. I will merge this and push a  request to main fork.
